### PR TITLE
Implement Async EMR container sensor async

### PR DESCRIPTION
**Pull request by bharanidharan14**
Story id : https://github.com/astronomer/astronomer-providers/issues/120
- Added Sensor, Hook, Trigger
- Added Test case for Hook, Trigger, sensors
- Added doc string, mypy fix

### CODE REVIEW

File: astronomer/providers/amazon/aws/sensors/emr.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    def execute_complete(self, context: Dict[Any, Any], event: Dict[str, str]) -> None:
+        """
+        Callback for when the trigger fires - returns immediately.
+        Relies on trigger to throw an exception, otherwise it assumes execution was
+        successful.
+        """
+        if event:
+            if event["status"] == "success":
+                return event["message"]
```
Comment (pankajastro) [inline astronomer/providers/amazon/aws/sensors/emr.py:46]: returning here may push value in xcom

Comment (pankajastro) [inline astronomer/providers/amazon/aws/sensors/emr.py:46]: returning here may push value in xcom

Comment (bharanidharan14) [reply astronomer/providers/amazon/aws/sensors/emr.py]: Fixed

File: astronomer/providers/amazon/aws/example_dags/example_emr.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import os
+from datetime import datetime, timedelta
+
+from airflow import DAG
+from airflow.providers.amazon.aws.operators.emr import EmrContainerOperator
+
+# [START howto_operator_emr_eks_env_variables]
+# from astronomer.providers.amazon.aws.sensors.emr import EmrContainerSensorAsync
```
Comment (kaxil) [inline astronomer/providers/amazon/aws/example_dags/example_emr.py:8]: ?

File: astronomer/providers/amazon/aws/sensors/emr.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    EmrContainerSensorAsync is async version of EmrContainerSensor,
+    Asks for the state of the job run until it reaches a failure state or success state.
+    If the job run fails, the task will fail.
+
+    :param virtual_cluster_id: Reference Emr cluster id
+    :param job_id: job_id to check the state
+    :param max_retries: Number of times to poll for query state before
+        returning the current state, defaults to None
+    :param aws_conn_id: aws connection to use, defaults to 'aws_default'
```
Comment (kaxil) [inline astronomer/providers/amazon/aws/sensors/emr.py:19]:
```suggestion
    :param aws_conn_id: aws connection to use, defaults to ``aws_default``
```

So they are rendered nicely as docstrings use RST

File: astronomer/providers/amazon/aws/triggers/emr.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+from typing import Any, AsyncIterator, Dict, Optional, Tuple
+
+from airflow.triggers.base import BaseTrigger, TriggerEvent
+
+from astronomer.providers.amazon.aws.hooks.emr import EmrContainerHookAsync
+
+
+class EmrContainerSensorTrigger(BaseTrigger):
+    """
+    EmrContainerSensorTrigger is fired as deferred class with params to run the task in trigger worker
```
Comment (kaxil) [inline astronomer/providers/amazon/aws/triggers/emr.py:11]: We should mention when this is fired, for example "This is fired when EMR container is created".

A good example is https://github.com/apache/airflow/blob/cb41d5c02e3c53a24f9dc148e45e696891c347c2/airflow/triggers/temporal.py#L26-L31

File: astronomer/providers/amazon/aws/sensors/emr.py (hunk 1 above)
Comment (phanikumv) [reply astronomer/providers/amazon/aws/sensors/emr.py]: what is done to resolve this comment @bharanidharan14 ?

Comment (bharanidharan14) [reply astronomer/providers/amazon/aws/sensors/emr.py]: Instead of retuning the success event, raised airflow exception when there is an error

**Review by phanikumv — COMMENTED** [review]
Can you rebase the branch from main?

File: astronomer/providers/amazon/aws/sensors/emr.py (hunk 3 above)
Comment (bharanidharan14) [reply astronomer/providers/amazon/aws/sensors/emr.py:19]: FIxed` :param aws_conn_id: aws connection to use, defaults to ``aws_default```

File: astronomer/providers/amazon/aws/example_dags/example_emr.py (hunk 2 above)
Comment (bharanidharan14) [reply astronomer/providers/amazon/aws/example_dags/example_emr.py:8]: Removed the commented duplicate import

File: astronomer/providers/amazon/aws/triggers/emr.py (hunk 4 above)
Comment (bharanidharan14) [reply astronomer/providers/amazon/aws/triggers/emr.py:11]: Fixed as 

```
EmrContainerSensorTrigger is triggered when EMR container is created, it polls for the Job container status.
    It is fired as deferred class with params to run the task in trigger worker
```

**Comment by phanikumv** [comment]
@bharanidharan14 Please mark as ready for review if it is so

File: astronomer/providers/amazon/aws/example_dags/example_emr.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+from datetime import datetime, timedelta
+
+from airflow import DAG
+from airflow.providers.amazon.aws.operators.emr import EmrContainerOperator
+
+from astronomer.providers.amazon.aws.sensors.emr import EmrContainerSensorAsync
+
+# [START howto_operator_emr_eks_env_variables]
+VIRTUAL_CLUSTER_ID = os.getenv("VIRTUAL_CLUSTER_ID", "astro-emr-cluster")
+AWS_CONN_ID = os.environ.get("ASTRO_AWS_CONN_ID", "aws_default")
```
Comment (phanikumv) [inline astronomer/providers/amazon/aws/example_dags/example_emr.py:11]: why not os.getenv for AWS_CON_ID as well?

**Comment by bharanidharan14** [comment]
> @bharanidharan14 Please mark as ready for review if it is so

Moved it ready for review

File: astronomer/providers/amazon/aws/hooks/emr.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+from typing import Any
+
+from botocore.exceptions import ClientError
+
+from astronomer.providers.amazon.aws.hooks.base_aws_async import AwsBaseHookAsync
+
+
+class EmrContainerHookAsync(AwsBaseHookAsync):
+    """
+    Async hook inherits AwsBaseHookAsync to interact with AWS EMR Virtual Cluster to run,
```
Comment (phanikumv) [inline astronomer/providers/amazon/aws/hooks/emr.py:10]: You may want to re word as "The EmrContainerHookAsync interacts with an AWS EMR EKS Virtual Cluster....." to make the docstring more meaningful

File: tests/amazon/aws/triggers/test_emr.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    assert len(task) == 1
+    msg = f"EMR Containers sensor failed {mock_status}"
+    assert TriggerEvent({"status": "error", "message": msg}) in task
+
+
+@pytest.mark.asyncio
+@mock.patch("astronomer.providers.amazon.aws.hooks.emr.EmrContainerHookAsync.check_job_status")
+async def test_emr_container_sensors_trigger_exception(mock_query_status):
+    """
+    Test RedshiftClusterTrigger with pause cluster failure
```
Comment (phanikumv) [inline tests/amazon/aws/triggers/test_emr.py:126]: This is not a test for Redshift cluster. Please correct it.

File: astronomer/providers/amazon/aws/triggers/emr.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+from typing import Any, AsyncIterator, Dict, Optional, Tuple
+
+from airflow.triggers.base import BaseTrigger, TriggerEvent
+
+from astronomer.providers.amazon.aws.hooks.emr import EmrContainerHookAsync
+
+
+class EmrContainerSensorTrigger(BaseTrigger):
+    """
+    EmrContainerSensorTrigger is triggered when EMR container is created, it polls for the Job container status.
```
Comment (phanikumv) [inline astronomer/providers/amazon/aws/triggers/emr.py:11]: what is meant by Job container status ?

File: tests/amazon/aws/triggers/test_emr.py (hunk 7 above)
Comment (bharanidharan14) [reply tests/amazon/aws/triggers/test_emr.py:126]: Fixed:- removed Redshift cluster.

File: astronomer/providers/amazon/aws/triggers/emr.py (hunk 8 above)
Comment (bharanidharan14) [reply astronomer/providers/amazon/aws/triggers/emr.py:11]: Fixed to 

```
EmrContainerSensorTrigger is triggered when EMR container is created, it polls for the AWS EMR EKS Job status.
    It is fired as deferred class with params to run the task in trigger worker
```

File: astronomer/providers/amazon/aws/example_dags/example_emr.py (hunk 5 above)
Comment (bharanidharan14) [reply astronomer/providers/amazon/aws/example_dags/example_emr.py:11]: Both works, now fixed to `os.getenv`

File: astronomer/providers/amazon/aws/hooks/emr.py (hunk 6 above)
Comment (bharanidharan14) [reply astronomer/providers/amazon/aws/hooks/emr.py:10]: Fixed: - 
```
The EmrContainerHookAsync interact with AWS EMR EKS Virtual Cluster
    to run, poll jobs and return job status
```

**Review by phanikumv — APPROVED** [review]

**Pull request merged**